mypy/checker.py

```diff
@@ -3115,3 +3120,23 @@ def push_class(self, info: TypeInfo) -> Iterator[None]:
 @contextmanager
 def nothing() -> Iterator[None]:
     yield
+
+
+def make_inferred_type_mismatch_note(msg: str, context: Context, subtype: Type,
+                                     supertype: Type, supertype_str: str) -> str:
+    '''The user does not expect an error if the inferred container type is the same as the return
+    type of a function and the argument type(s) are a subtype of the argument type(s) of the
+    return type. This note suggests that they add a type annotation with the return type instead
+    of relying on the inferred type.
+    '''
+    if (msg.startswith(messages.INCOMPATIBLE_RETURN_VALUE_TYPE) and
+      isinstance(subtype, Instance) and isinstance(supertype, Instance) and
+      subtype.type.fullname() == supertype.type.fullname() and
+      subtype.args and supertype.args and
+      isinstance(context, ReturnStmt) and isinstance(context.expr, NameExpr) and
+      isinstance(context.expr.node, Var) and context.expr.node.is_inferred):
+        for subtype_arg, supertype_arg in zip(subtype.args, supertype.args):
+            if not is_subtype(subtype_arg, supertype_arg):
+                return ''
+        return 'Perhaps a type annotation is missing here? Suggestion: ' + supertype_str
```

Comment:
Might be better to also mention the name of the variable to make it clear that the proposed annotation is for the variable, not for the return statement. For example, `Perhaps you need a type annotation for "x"? Suggestion: List[float]`